### DISCUSSION TITLE
How do I avoid generating references  doing setlocation

### DISCUSSION THREAD

**Discussion by thyis** [fougue/mayo#215] [Show and tell]
1.  i chose a shell  moving  from original location to result location, 
![move](https://github.com/fougue/mayo/assets/63134171/585250a7-1b91-4c46-a3f8-fa03096bc926)
2.  the code :shapeTool->SetLocation(gfxLabel, TopLoc_Location(result), theRefLabel);
3.  export .stp
4.  import .stp
5.  it generating references, like  0:1:1:1,  How do I avoid generating [references.]
![result](https://github.com/fougue/mayo/assets/63134171/935bee75-6ad4-450e-bcd0-050ca7f37c72)

**Comment by HuguesDelorme** [comment]
`XCAFDoc_ShapeTool::SetLocation()` creates an instance(reference) if the input shape  label isn't already one.
For your specific case you have to detect if the selected assembly item is a reference or not(let's call it a "product" in that case).
Use function `XCaf::isShapeReference()` for that purpose.
If it's already a reference then call `XCAFDoc_ShapeTool::SetLocation()` as you do already.
Otherwise that's a product so you have to transform the `TopoDS_Shape` object itself. Use `BRepBuilderAPI_Transform` for that.
At the end update the shape in the assembly with `XCAFDoc_ShapeTool::SetShape()`

**Reply by thyis** [reply]
I am very grateful to you for your advice.  it work well for moving a product.

**Comment by thyis** [comment]
it is ok product . 
in the case :  i move one bolt , 
![a](https://github.com/fougue/mayo/assets/63134171/e6edc889-94a6-4ab4-8bcc-f808494d02df)
the code is:
if (XCAFDoc_ShapeTool::IsReference(gfxLabel))
{
	TDF_Label labelRef;
	if (XCAFDoc_ShapeTool::GetReferredShape(gfxLabel, labelRef))
	{
		shapeTool->SetLocation(gfxLabel, reftrsf, labelRef);
	}
}
then  it move all bolt.  What I want is to move only the selected object.
![b](https://github.com/fougue/mayo/assets/63134171/f764cad8-84f5-45f2-b3a5-ae04aed1c886)

**Reply by HuguesDelorme** [reply]
In that case you should not move the product but the instance(reference)
Your code is wrong, it should rather be like this:
````c++
TDF_Label labelProduct;
if (XCAFDoc_ShapeTool::IsReference(gfxLabel)) {
    TDF_Label labelRef;
    shapeTool->SetLocation(gfxLabel, reftrsf, labelRef/*or gfxLabel*/);
}
else if (XCAFDoc_ShapeTool::GetReferredShape(gfxLabel, labelProduct)) {
    // Use BRepBuilderAPI_Transform to transform the product shape location
    // Call XCAFDoc_ShapeTool::SetShape(...) on labelProduct with the updated shape
}
````

**Comment by thyis** [comment]
thanks , I still haven't been able to move one  bolt alone , i need moretime to study mayo.

**Comment by thyis** [comment]
If possible, coudld you please kindly help to provide some details regarding how geometries can be undo/ redo in mayo ? This would be helpful and flexible for some simple geometry creatations and operations .
